File: android/app/src/main/java/com/noop/testcentre/TestCentre.kt
package com.noop.testcentre

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONObject

/**
 * The Test Centre orchestration surface (Kotlin twin of the Swift TestCentre).
 *
 * Backed by a SINGLE SharedPreferences file "noop_testcentre" for all NEW Test Centre flags,
 * consolidating what used to be scattered across noop_prefs / noop_experiments / noop_debug_export and
 * ~10 non-reactive mutableStateOf mirrors in SettingsScreen. The legacy experiment + debug-export keys
 * are PRESERVED in their own files (read through their existing accessors), so migration never loses a
 * setting (spec section 10). [active] is the zero-cost gate engines check before emitting a tagged line.
 *
 * The constructor takes a [SharedPreferences] directly so the surface is exercisable on the plain JVM
 * (no Robolectric, matching the rest of the suite, see DeviceRegistryTest / MoodStoreTest); production
 * uses [from] to bind the real "noop_testcentre" file.
 */
class TestCentre internal constructor(private val prefs: SharedPreferences) {

    /** The zero-cost gate. `if (testCentre.active(TestDomain.SLEEP)) client.log(line, TestDomain.SLEEP)`. */
    fun active(d: TestDomain): Boolean {
        if (d == TestDomain.UNIVERSAL) return anyActive()        // universal rides whatever mode is on
        if (prefs.getBoolean(ACTIVE_PREFIX + TestDomain.MASTER.id, false)) return true  // master = all on
        return prefs.getBoolean(ACTIVE_PREFIX + d.id, false)
    }

    /** True when ANY non-universal mode is on (drives the testing-on banner plus the universal traces). */
    fun anyActive(): Boolean = TestDomain.values().any {
        it != TestDomain.UNIVERSAL && prefs.getBoolean(ACTIVE_PREFIX + it.id, false)
    }

    fun activate(d: TestDomain) {
        prefs.edit()
            .putBoolean(ACTIVE_PREFIX + d.id, true)
            .putLong(STARTED_PREFIX + d.id, System.currentTimeMillis() / 1000L)
            .apply()
    }

    fun deactivate(d: TestDomain) {
        prefs.edit().putBoolean(ACTIVE_PREFIX + d.id, false).apply()
    }

    /** Unix seconds the mode was last activated, or null if never. */
    fun startedAt(d: TestDomain): Long? {
        val t = prefs.getLong(STARTED_PREFIX + d.id, 0L)
        return if (t > 0L) t else null
    }

    fun answers(d: TestDomain): Map<String, String> {
        val raw = prefs.getString(ANSWERS_PREFIX + d.id, null) ?: return emptyMap()
        return runCatching {
            val o = JSONObject(raw)
            buildMap { for (k in o.keys()) put(k, o.getString(k)) }
        }.getOrDefault(emptyMap())
    }

    fun setAnswers(d: TestDomain, m: Map<String, String>) {
        val o = JSONObject()
        for ((k, v) in m) o.put(k, v)
        prefs.edit().putString(ANSWERS_PREFIX + d.id, o.toString()).apply()
    }

    /**
     * One-time migration. Idempotent, guarded by the v1 bool. Phase 1 has no domain-activation state to
     * seed from the legacy toggles (those are advanced experimental flags, gathered by the IA but not
     * domain activations), so this only stamps the guard. The legacy PuffinExperiment (noop_experiments)
     * and DebugExportSettings (noop_debug_export) keys are read in place through their existing
     * accessors; nothing is moved or renamed.
     */
    fun migrate() {
        if (prefs.getBoolean(MIGRATED_KEY, false)) return
        prefs.edit().putBoolean(MIGRATED_KEY, true).apply()
    }

    companion object {
        private const val PREFS = "noop_testcentre"
        private const val ACTIVE_PREFIX = "testcentre.active."
        private const val STARTED_PREFIX = "testcentre.startedAt."
        private const val ANSWERS_PREFIX = "testcentre.answers."
        private const val MIGRATED_KEY = "testcentre.migrated.v1"

        fun from(context: Context): TestCentre =
            TestCentre(context.getSharedPreferences(PREFS, Context.MODE_PRIVATE))
    }
}

File: Strand/System/TestCentre.swift
import Foundation
import StrandAnalytics

/// The Test Centre orchestration surface: per-domain activation, a SINGLE consolidated prefs namespace,
/// and a one-time read-through migration that gathers the scattered legacy keys behind one accessor
/// WITHOUT renaming them (spec section 10). `active(_:)` is the zero-cost gate engines check before
/// emitting a tagged line, so an emitter on the GATT/analytics queue pays one Bool read when a mode is
/// off. The Kotlin twin is TestCentre.kt, backed by a single "noop_testcentre" SharedPreferences file.
public enum TestCentre {

    // SINGLE namespace for all NEW Test Centre flags. The migrated LEGACY keys keep their original
    // names (see migrate()), so no user loses a setting.
    private static let activePrefix = "testcentre.active."          // + domain.id  -> Bool
    private static let startedPrefix = "testcentre.startedAt."      // + domain.id  -> Double (unix)
    private static let guidedTargetPrefix = "testcentre.target."    // + domain.id  -> Int (nights/days)
    private static let answersPrefix = "testcentre.answers."        // + domain.id  -> [String:String] (Data)
    private static let migratedKey = "testcentre.migrated.v1"

    private static let master = TestDomain.master

    /// The zero-cost gate. `if TestCentre.active(.sleep) { live.append(log:domain:) }`. `nonisolated`
    /// plus a single Bool read so an emitter on any actor pays almost nothing when the mode is off.
    public nonisolated static func active(_ d: TestDomain) -> Bool {
        if d == .universal { return anyActive }      // universal rides whatever mode is on
        if UserDefaults.standard.bool(forKey: activePrefix + master.id) { return true }  // master = all on
        return UserDefaults.standard.bool(forKey: activePrefix + d.id)
    }

    /// True when ANY non-universal mode is on (drives the "testing on" banner plus the universal traces).
    public nonisolated static var anyActive: Bool {
        TestDomain.allCases.contains {
            $0 != .universal && UserDefaults.standard.bool(forKey: activePrefix + $0.id)
        }
    }

    @MainActor public static func activate(_ d: TestDomain) {
        UserDefaults.standard.set(true, forKey: activePrefix + d.id)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: startedPrefix + d.id)
    }

    @MainActor public static func deactivate(_ d: TestDomain) {
        UserDefaults.standard.set(false, forKey: activePrefix + d.id)
    }

    public static func startedAt(_ d: TestDomain) -> Date? {
        let t = UserDefaults.standard.double(forKey: startedPrefix + d.id)
        return t > 0 ? Date(timeIntervalSince1970: t) : nil
    }

    /// The guided capture target (nights for Sleep, days for Battery), 0 when unset. Read-through of
    /// the single namespace; the guided flow (Group E/F) writes it via `setGuidedTarget`.
    public static func guidedTarget(_ d: TestDomain) -> Int {
        UserDefaults.standard.integer(forKey: guidedTargetPrefix + d.id)
    }

    @MainActor public static func setGuidedTarget(_ count: Int, for d: TestDomain) {
        UserDefaults.standard.set(count, forKey: guidedTargetPrefix + d.id)
    }

    public static func answers(_ d: TestDomain) -> [String: String] {
        guard let data = UserDefaults.standard.data(forKey: answersPrefix + d.id),
              let m = try? JSONDecoder().decode([String: String].self, from: data) else { return [:] }
        return m
    }

    @MainActor public static func setAnswers(_ m: [String: String], for d: TestDomain) {
        if let data = try? JSONEncoder().encode(m) {
            UserDefaults.standard.set(data, forKey: answersPrefix + d.id)
        }
    }

    /// One-time migration: fold the scattered @AppStorage / PuffinExperiment / ScheduledDebugExport keys
    /// behind this surface WITHOUT renaming them (read-through). Existing keys are PRESERVED (spec section
    /// 10): the experimental toggles keep their PuffinExperiment.*Key names, the scheduled export keeps its
    /// debugExport.* names. This only seeds the NEW testcentre.* surface and never deletes a legacy key.
    /// Idempotent, guarded by the migratedKey bool.
    @MainActor public static func migrate() {
        guard !UserDefaults.standard.bool(forKey: migratedKey) else { return }
        // Phase 1 has no testcentre.active.* state to seed from the legacy toggles (those are advanced
        // experimental flags, gathered by the IA but not domain activations), so the migration only
        // stamps the guard. The legacy keys are read in place through their existing accessors:
        //   PuffinExperiment.defaultsKey / .deepDataKey / .broadcastHrKey / .keepRealtimeForDataKey /
        //   .experimentalSleepV2Key / .autoDetectWorkoutsKey, and the ScheduledDebugExport "debugExport.*"
        //   keys. Nothing is moved; the Test Centre screen reads them where they already live.
        UserDefaults.standard.set(true, forKey: migratedKey)
    }
}
